aten/src/ATen/native/quantized/cpu/qnnpack_utils.h

```diff
@@ -2,10 +2,24 @@
 
 #ifdef USE_QNNPACK
 #include <qnnpack.h>
+#ifdef C10_MOBILE
+#include "caffe2/utils/threadpool/ThreadPool.h"
+#endif
 
 struct QnnpackOperatorDeleter {
   void operator()(qnnp_operator_t op) {
     qnnp_delete_operator(op);
   }
 };
+
+class ThreadPoolMobile {
+public:
+  static pthreadpool_t qnnpack_threadpool();
+#ifdef C10_MOBILE
+private:
+  static std::unique_ptr<caffe2::ThreadPool> thread_pool_;
```

Comment:
Just FYI, because of the global lock `caffe2::ThreadPool::execution_mutex_`, this means two independent models executing convolutions on separate threads will contend on this shared resource. Not sure if this is a problem for you folks.

Reply:
Thanks, I don't think this will affect our use case for now, but will keep it in mind. cc @ljk53

Reply:
Not sure if pytorch has something equivalent to c2's workspace - ideally these global resources should be allocated and owned by some global context object.

Reply:
@ljk53 yes, the concept of "model" or "session" owned resources is a useful one that pytorch doesn't have I believe.

Reply:
cc @ilia-cher - have you thought about threading resource management API for devices? (for server it's probably fine as it's usually forever running binary anyway)

Reply:
As for "shared" thread pool - I'd say it's a good idea to have a single one. Is there any case we'd want to create thread pool per model?

Regarding clear up - it might be a good idea to clean up resources when the last Interpreter is destroyed assuming that there's no python extension present (though it still misses the case of C++ API just calling ops on tensors). Alternatively we can just expose a global call to "garbage collect resources" that tries to reset threading libraries.

What is the preferred strategy on device in general? if the last model is destroyed - clean up the thread pool?

Reply:
For the "shared" thread pool part - I agree it's bounded by cpu cores anyway unless we carefully config thread pool size for each model. But having a workspace kinda concept to hold all global resources is probably good for "clearing up" purpose? Right now I've seen static / thread-local scattered in the codebase. We might want to clean up at some point?

Yeah, we can associate resource clean up with the life-cycle of interpreter/predictor instance we expose to the application. My question is more about "at::parallel" level API to expose resource management functionality in an abstract way as "at::parallel" primitives hide thread-pool (or whatever implementation) as implementation details. Similarly - what if user wants to create and warm up thread-pool before first inference?